create-event-detectors.js

```diff
@@ -0,0 +1,549 @@
+const ModuleRegistry = Java.type('com.serotonin.m2m2.module.ModuleRegistry');
+const AlarmLevels = Java.type('com.serotonin.m2m2.rt.event.AlarmLevels');
+const dataPointService = services.dataPointService;
+const eventHandlerService = services.eventHandlerService;
+const EventHandlerDao = Java.type('com.serotonin.m2m2.db.dao.EventHandlerDao');
+const ArrayList = Java.type('java.util.ArrayList');
+
+
+
+const mainHeaders = [];
+mainHeaders.push("dataPointId", "dataPointXid", "detectorType", "detectorName", "alarmLevel", "limit", "stateValues", "stateInverted", "duration", "durationType", "handlers_to_link", "dataPointName", "dataPointType");
+
+const compare = (actual, expected, message) => {
+    if (actual === expected) {
+        return
+    }
+    else
+    {
+        if (actual.toUpperCase() === expected.toUpperCase()) {
+            verbose(`WARNING: Exact case match failed for ${actual} and ${expected}. Tested with toUpperCase() and passed validation.`);
+            return
+        }
+        else
+        throw new Error(`${message}: ${actual} is NOT equal to ${expected}!`)
+    }
+}
+
+function readCsv(fileStore, filePath) {
+    const path = services.fileStoreService.getPathForRead(fileStore, filePath);
+    const lines = Array.from(Files.readAllLines(path));
+    const header = lines.shift().replace(/["']/g, '').replace(/\s+/g, "").split(',');
+
+    //Validate headers 
+    for (let index = 0; index < mainHeaders.length; index++) {
+        compare(header[index], mainHeaders[index], `Fatal Error. Expected column header name ${mainHeaders[index]} mismatch`);
+        //If this is an upper case match, we need to update the mainHeaders array
+        if (header[index].toUpperCase() === mainHeaders[index].toUpperCase()) {
+            header[index] = mainHeaders[index];
+        }
+    }
+
+    //Get the data
+    for (let i = 0; i < lines.length; i++) {
+        const row = lines[i].replace(/["']/g, "").split(',');
+        const data = lines[i] = {};
+        for (let j = 0; j < row.length; j++) {
+            data[header[j]] = row[j];
+        }
+    }
+    return lines;
+}
+
+function assureNotEmpty(name, value) {
+    if (!value) {
+        throw new Error(`Missing required value: ${name}. Validation failed.`);
+    }
+    else if (value == RESERVED_EMPTY) {
+        throw new Error(`Missing required value: ${name}. Value set to ${RESERVED_EMPTY}. Validation failed.`);
+    }
+    return;
+}
+
+const isTypeSupported = (type) => {
+    switch (type) {
+        case "HIGH_LIMIT":
+        case "LOW_LIMIT":
+        case "MULTISTATE_STATE":
+        case "BINARY_STATE":
+        case "NO_UPDATE":
+        case "UPDATE":
+            verbose(`${type}: supported`);
+            return;
+        default:
+            if (type) throw new Error(`detectorType ${type}: Not Supported!`);
+            else throw new Error(`: No supported detectorType provided!`);
+    }
+}
+
+const verbose = (logMessage) => {

```

Comment:
I'm surprised calling this lambda works above this line. Functions are hoisted to the top of the file, a `const` is not.

Reply:
Actually I guess this makes sense, it is in scope when it is called.

Reply:
That was a carry-over from the original script. I thought it was a bit odd, but I left it in. I'll re-write this as a simple function to make it more readable.